rm(list=ls()) # clear workspace 


folder_output = "output_test" 
folder_tmp = "tmp_revision"
folder_data = "data"

b.write_paper = FALSE


source("config.R")
v.partitions <- c("promoter_5kb", "promoter_1kb", "gene", "five_prime_UTR", "exon", "intron", "three_prime_UTR", "post_gene_1kb", "non_genic") 


source("utils.R")

## Extract condition
require(reshape2)
require(ggplot2)
library(dplyr)
require(seqinr)
require(Biostrings)
require(VariantAnnotation)
require(foreach)
require(doParallel)
require(ggplot2)
require(BSgenome)
require(VariantAnnotation)
require(plotly)
require(rtracklayer)
library(stringr)




# datasets
path.bQTLs <- "data/revision/B73.Mo17.WW.PF.uni_EG80.GT.RN.BZR1.GEM.csv" # old
#path.bQTLs <- "data/revision/bzr1_counts/B73.Mo17.WW.PF.uni_EG80.GT.RN.BZR1.GEM.csv"
path.peaks = "data/revision/BZR1_6.GEM_events.br.filtered.1_4_2_5_3_6.bed" # pre-filtered peaks # BZR1_6.GEM_events.bed"
# for testing purposes of peak integration
# v.paths.peaks.replicates = c(path.peaks, path.peaks, path.peaks, path.peaks, path.peaks, path.peaks)

path.genomic_sequences = "data/revision/ref_B73Mo17.fasta"


path.gene_annotation <- "data/revision/B73_Mo17_CSHL.gff"

path.phenotypic_snps <- "data/revision/gwas/remapped_AGPv5_chr_uplifted_Wallace_etal_2014_PLoSGenet_GWAS_hits.bed"

path.raw_input_reads <-"data/revision/INPUT.q255.RPGC.exScal.bin1.sm0.bw"
#path.raw_input_reads <- "data/revision/BZR7_12.RPGCq255.b1.ct.bigwig" # old
path.enhancer <- "data/revision/remapped_B73_enhancers_v5.bed"

path.expression <- "data/revision/expression/Baldauf_2016_dataset.txt"

#### START PEAK PREPROCESSING
if(FALSE){
  
  message("Validate called peaks.")
  
  source("utils/binding_peaks_revision.R")
  merge_peaks(path.peaks,
               v.paths.peaks.replicates,
               i.ref=1, # reference file 
               th.min_number = 5, # min number of add. files to find peak found in ref. file
               folder = folder_tmp)
  
  filter_peaks_w_merged_replicaties(path.peaks, 
                                    th.dist = 500, # max allowed distance between peaks in diff. files 
                                    folder = folder_tmp)
  
  # TODO: check WARNING 
  #1: In if (dist < th.dist) { ... :
  #    the condition has length > 1 and only the first element will be used
  #  2: In if (dist < th.dist) { ... :
  
  gc()
  
}
#### END PEAK PREPROCESSING


# TODO: put into separate function pre-processing !!!

# The data is generated from BZR1 ChIP in all six replicates in B73xMo17, mapped to a diploid genome, i.e. the two parental genomes pasted together. 
# Only unique mapping reads were kept, which means that only reads with sequence differences between the two inbred lines are mapped, 
# identical regions in the genome have no reads. However, since the ASB analysis only concerns SNPs that does not affect this part of the analysis. 
# The SNPs were generated by whole genome alignment, SNPs that map to several places in one of the parental genomes were excluded.
# That leaves us with 10.973.994 SNPs.
df.snps <- read.table(path.bQTLs,  sep ="\t", stringsAsFactors = FALSE)
names(df.snps) <- c("B73-chr", "B73-pos", "B73-base", "Mo17-base", "Mo17-position", "Genotype",
                      "CPM-B73", "CPM-Mo17", "Peak-B73", "Peak-Mo17", "post-frequency", "Reads-B73", "Reads-Mo17")

message(nrow(df.snps), " SNPs from BZR1 ChIP in all six replicates in B73xMo17")

df.snps["refCount"] <- df.snps["Reads-B73"]
df.snps["altCount"] <- df.snps["Reads-Mo17"]
df.snps["totalCount"] <- df.snps$refCount + df.snps$altCount
df.snps["refAllele"] <- df.snps["B73-base"]
df.snps["altAllele"] <- df.snps["Mo17-base"]

minReadsPerAllele = 1 # only work with covered snps
df.snps <- subset(df.snps, df.snps$refCount >= minReadsPerAllele & df.snps$altCount >= minReadsPerAllele)

df.snps["Mo17-chr"] <- as.character(unlist(lapply(strsplit(df.snps$`Mo17-position`, ":"), function(m) m[[1]])))
df.snps["Mo17-pos"] <- as.numeric(unlist(lapply(strsplit(df.snps$`Mo17-position`, ":"), function(m) m[[2]])))

# conservative filtering (only B73 / Mo17 chromosomes 1 to 10)
df.snps <- subset(df.snps, df.snps$`B73-chr` %in% paste("B73-chr", 1:10, sep = "") & df.snps$`Mo17-chr` %in% paste("Mo17-chr", 1:10, sep = "")) # hard filter on removing all scaffolds - show table

print(table(df.snps$`B73-chr`))
print(table(df.snps$`Mo17-chr`))

df.snps["POSTfreq"] <- df.snps$refCount / df.snps$totalCount
df.snps["POSTallele"] <- ifelse(df.snps$altCount > df.snps$refCount, df.snps$altAllele, df.snps$refAllele)

#### Select background candidates #### 

maxReadsPerAllele = 15
maxReadDepth = 20

df.bgSNPs.candidates <- subset(df.snps, 
                               df.snps$refCount <= maxReadsPerAllele & df.snps$altCount <= maxReadsPerAllele)
df.bgSNPs.candidates <- subset(df.bgSNPs.candidates, df.bgSNPs.candidates$totalCount <= maxReadDepth)  
df.bgSNPs.candidates <- subset(df.bgSNPs.candidates, df.bgSNPs.candidates$POSTfreq != 0.5) # This ignores bg SNPS with allelic bias == 0.5 !!!!

# message("background SNP candidates after min reads per allele filter: ", nrow(df.bgSNPs.candidates))
message(nrow(df.bgSNPs.candidates), " non-significant SNPs located outside high-confidence BZR1 binding peaks")
hist(df.bgSNPs.candidates$POSTfreq, breaks = 100, main = paste("Allelic bias of", nrow(df.bgSNPs.candidates), "non-significant (background) SNPs located"), xlab = "allelic bias", cex.main = 0.7)


#### Select putative bQTLs ####
minReadDepth = 25
df.snps <- subset(df.snps, df.snps$refCount >= minReadDepth | df.snps$altCount >= minReadDepth)
message("input bQTL after min reads per allele filter: ", nrow(df.snps))
message(nrow(df.snps), " SNPs from BZR1 ChIP in all six replicates in B73xMo17 (10 chromosomes only)")
hist(df.snps$POSTfreq, breaks = 100, main = paste("Allelic bias of", nrow(df.snps), " SNPs from BZR1 ChIP in all six replicates in B73xMo17 (10 chromosomes only)"), xlab = "allelic bias", cex.main = 0.7)


prob.bias <- median(df.snps$refCount / df.snps$totalCount)
message("bias probability (median): ", prob.bias)
th.p.bQTL = 0.001
fdr <- p.adjust(sapply(1:nrow(df.snps), function(i) binom.test(as.integer(df.snps$altCount[i]), as.integer(df.snps$totalCount[i]), p = 1 - prob.bias)$p.value), "bonferroni")
df.snps["p-value (corrected)"] <- fdr

# Of the remaining 429236 SNPs, we determined significant variation of median allele
# frequency of 0.497 using a binomial test with a p-value cutoff of â‰¤ 0.001 adjusted for
# multiple testing using Bonferroni correction (n=46286 SNPs)
df.bQTLs <- subset(df.snps, df.snps$`p-value (corrected)` < th.p.bQTL)
df.bQTLs <- subset(df.bQTLs, df.bQTLs$POSTfreq < 1 & df.bQTLs$POSTfreq > 0)

message(nrow(df.bQTLs), " bQTLs (significance p < ", th.p.bQTL, " )")
hist(df.bQTLs$POSTfreq, breaks = 100, main = paste("Allelic bias of", nrow(df.bQTLs), "putative bQTLs (p < 0.001)"), xlab = "allelic bias", cex.main = 0.7)
# write.csv(df.bQTLs, "57414_bQTLs.csv", quote = F, row.names = F)
# saveRDS(df.bQTLs, paste(folder_tmp, "57414_bQTLs.rds", sep = "/"))

message("Estimate significant bQTLs in binding peaks.")

source("utils/binding_peaks_revision.R")

df.peaks <- load_peaks(path.peaks)  # original merged peak data (without replicated-specific filtering)

if(FALSE){
  
  source("utils/genomic_location_revision.R")
  
  df.peaks_genomic_location <- add_genomic_location(df.peaks, 
                                                   chr = "seqnames",
                                                   pos = "center",
                                                   df.gene_annotation,
                                                   n.cpus = 5)

  write.csv(df.peaks_genomic_location, "52765_peaks_w_genomic_location_w_gene_association.csv", quote = F, row.names = F)

}



df.bQTLs <- bQTLs_in_peaks(df.bQTLs, df.peaks)

message(nrow(df.bQTLs), " SNPs located within high-confidence BZR1 binding peaks (all 6 replicated merged)")
hist(df.bQTLs$POSTfreq, breaks = 100, main = paste("Allelic bias of", nrow(df.bQTLs), "bQTL located within high-confidence BZR1 binding peaks"), 
     xlab = "allelic bias", cex.main = 0.7)

write.csv(df.bQTLs, "33267_bQTLs_in_peaks.csv", quote = F, row.names = F)
#write.table(write_bed(df.bQTLs), "33267_bQTLs_in_peaks.bed", row.names = F, col.names = F, quote = F, sep = "\t")
# saveRDS(df.bQTLs, paste(folder_tmp, "33267_bQTLs_in_peaks.rds", sep = "/"))


source("utils/linkage_disequilibrium.R")
df.ASBs <- linkage_disequlilibrium(df.bQTLs,
                                   df.peaks,
                                   mode = "peak_distance",
                                   s.disequilibriumDistance = 75,
                                   n.chromosomes = 10)

message(nrow(df.ASBs), " ASBs (bQTLs in peaks after linkage disequilibrium)")
bQTL_scatterplot(df.ASBs)
hist(df.ASBs$POSTfreq, breaks = 100, main = paste("Allelic bias of", nrow(df.ASBs), "ASBs after linkage disequilibrium (min peak distance)"), 
     xlab = "allelic bias", cex.main = 0.7)

write.csv(df.ASBs, "7817_ASBs_75bp_min_peak_distance.csv", quote = F, row.names = F)
write.table(write_bed(df.ASBs), "7817_ASBs_75bp_min_peak_distance.bed", row.names = F, col.names = F, quote = F, sep = "\t")

asb_environment <- extract_sequences(df.ASBs,
                                     genome_sequences = readDNAStringSet(path.genomic_sequences),
                                     s.env_bps = 100,
                                     n.chromosomes = 10)

writeXStringSet(asb_environment, "200_bp_environment_sequences_from_7817_ASBs_75bp_min_peak_distance.fasta")





message("Identify and filter (blacklist) putative ASBs explained by systematic bias in genomic region from input read data")


source("utils/blacklisting.R")

res <- identify_systematic_bias(df.ASBs, 
                                df.bgSNPs.candidates,
                                path.raw_input_reads,
                                n.snps.split = 50,
                                th.bp_ranges = 1000,
                                n.bg.multiplier = 5,
                                bQTL_only=F)

#saveRDS(res, paste(folder_tmp, "res_7817_ASBs_w_bgSNPs.rds", sep = "/"))
res <- readRDS(paste(folder_tmp, "res_7817_ASBs_w_bgSNPs.rds", sep = "/"))

df.ASBs <- filter_systematic_bias(df.bQTLs = res$df.bQTLs,
                                  v.bg.genomic_postfrequency = res$df.bgSNPs$pf.region,
                                  th.prob = 0.05, 
                                  plot.trendline = F)

message(nrow(df.ASBs), " blacklist filtered ASBs (bQTLs in peaks after linkage disequilibrium)")
hist(df.ASBs$POSTfreq, breaks = 100, main = paste("Allelic bias of", nrow(df.ASBs), "ASBs after linkage disequilibrium (min peak distance)"), 
     xlab = "allelic bias", cex.main = 0.7)

write.csv(df.ASBs, "6143_ASBs_75bp_min_peak_distance_blacklist.csv", quote = F, row.names = F)
write.table(write_bed(df.ASBs), "6143_ASBs_75bp_min_peak_distance_blacklist.bed", row.names = F, col.names = F, quote = F, sep = "\t")

asb_environment <- extract_sequences(df.ASBs,
                                     genome_sequences = readDNAStringSet(path.genomic_sequences),
                                     s.env_bps = 100,
                                     n.chromosomes = 10)

writeXStringSet(asb_environment, "200_bp_environment_sequences_from_6143_ASBs_75bp_min_peak_distance.fasta")



message("perform gene partitioning")
df.gene_annotation <- load_gene_annotation(path.gene_annotation)


source("utils/genomic_location_revision.R")
df.ASBs_genomic_location = add_genomic_location_bQTLs(df.ASBs,  
                                                      df.gene_annotation,
                                                      n.cpus = 5) 
                              
#saveRDS(df.ASBs_genomic_location, paste(folder_tmp, "6143_ASBs_w_genomic_location.rds", sep = "/"))                  
message(nrow(df.ASBs_genomic_location), " high confidence allele-specific BZR1 binding sites located near ", length(unique(df.ASBs_genomic_location$gene.ID)), " flanking genes")
write.csv(df.ASBs_genomic_location, "6143_ASBs_w_genomic_location.csv", quote = F, row.names = F)

df.distribution <- data.frame(ASBs = apply(df.ASBs_genomic_location[,v.partitions], 2, table)["yes",])
df.distribution["%ASBs"] <-  round(df.distribution$ASBs / sum(df.distribution$ASBs) * 100, 1)

message("Distribution of ASBs positions around / withing genes:")
print(df.distribution)


### BACKGROUND 

message("Control background SNP (bgSNP) sampling")
# Functional GWAS variants have been shown to be significantly enriched in gene proximal
# regions. Therefore, control background SNPs (bgSNPs) were proportionally sampled (excluding ASBs) 
# per chromosome and genomic location (i.e. 5 - 1kb upstream, 1kb upstream - TSS, 5â€™UTR, exon, intron, 3â€™UTR, TTS - 1 kb downstream, intergenic) to match
# the genomic distribution of the ASBs dataset. Additionally, we checked that ASBs and bgSNPs showed a similar minor allele-frequency . In total,
# 168950 bgSNPs were sampled, yielding approximately 50 times as many background SNPs per genome location, compared to the number of ASBs within each location
df.snps_genomic_location = add_genomic_location_bQTLs(df.bgSNPs.candidates,  
                                                      df.gene_annotation,
                                                      n.cpus = 2) 

# saveRDS(df.snps_genomic_location, paste(folder_tmp, "6181870_bgSNPs_candidates_w_genomic_location.rds", sep = "/"))


source("utils/qtl_background_sampling.R")

v.partitions <- c("promoter_5kb", "promoter_1kb", "gene", "five_prime_UTR", "exon", "intron", "three_prime_UTR", "post_gene_1kb", "non_genic") 

res <- create_background_QTLs(df.ASBs_genomic_location,
                              df.snps_genomic_location, 
                              n.bg.multiplier = 40,  # 50
                              v.partitions = v.partitions, 
                              seed = 1234)
                                    
df.bgSNPs <- res$df.bgSNPs
message(nrow(df.bgSNPs), " sampled, non-significant (background) SNPs located outside high-confidence BZR1 binding peaks (all 6 replicated merged)")
hist(df.bgSNPs$POSTfreq, breaks = 100, main = paste("Allelic bias of", nrow(df.bgSNPs), "sampled, non-significant (background) SNPs located outside high-confidence BZR1 binding peaks"), 
     xlab = "allelic bias", cex.main = 0.7)


genomic_distribution(df.ASBs_genomic_location, df.bgSNPs, v.partitions)


bQTL_scatterplot(df.bgSNPs)
# distribution - scatter
# saveRDS(df.bgSNPs, paste(folder_tmp, "317094_bgSNPs_w_genomic_location.rds", sep = "/"))
# write.csv(df.bgSNPs, "317094_background_SNPs_w_genomic_location.csv", quote = F, row.names = F)
# write.table(write_bed(df.bgSNPs), "317094_background_SNPs.bed", row.names = F, col.names = F, quote = F, sep = "\t")

### GWAS ###

message("-------------- GWAS ------------- ")


# Fig. 4. ASBs of ZmBZR1 are linked to growth and disease related traits. 
# b) Association of ASBs with the curated 4041 significant GWAS hits for selected phenotypes of
# the NAM population. Abbreviations: Av: Average; intern: internode; len: length; w. pl.: whole plant;
# b.: below c) VCAP Variance component analysis results. Variance explained (h2) by the ASB-
#   SNP set (bars) and background SNP set (violin plots, derived from the permutation results).
# 29Orange color in the bars denotes a significantly higher variance explained (h2) by ASBs than
# expected by chance (one-sided permutation test < 0.1).

df.phenotypic_snps <- read.table(path.phenotypic_snps)
names(df.phenotypic_snps) <- c("chr", "pos.start", "pos.end", "trait")
v.traits  <- unique(df.phenotypic_snps$trait)
print(table(df.phenotypic_snps$trait))


source("utils/gwas_enrichment_revision.R")

res <- perform_gwas(df.ASBs, df.bgSNPs, B73.only = F, th.pval = 0.05)
write.csv(res$d.gsea, "2000bp_dist_phenotypic_gwas_6143_ASBs_317094_bgSNPs.csv",  row.names = FALSE, quote = FALSE)

write.csv(res$res.ASBs$df.bQTLs_w_phenotypes, "2000bp_dist_phenotypic_gwas_6143_ASBs_w_traits.csv",  row.names = FALSE, quote = FALSE)

df.bQTLs_w_phenotypes <- res$res.ASBs$df.bQTLs_w_phenotypes
df.bgSNPs_w_phenotypes <- res$res.bgSNPs$df.bQTLs_w_phenotypes
v.traits <- names(res$res.ASBs$l.number_per_trait)

n <- length(v.traits) * length(v.partitions)
d.gsea <- data.frame(trait = rep("", n), 
                     partition = rep("", n),
                     p.val = rep(1, n),
                     number = rep(0, n), 
                     percent = rep(0, n), 
                     foldchange = rep(1, n), stringsAsFactors = FALSE)


idx <- 1
for(trait in v.traits){
  
  df.bQTLs.trait <- subset(df.bQTLs_w_phenotypes, df.bQTLs_w_phenotypes[,trait] == "yes")
  df.bgSNPs.trait <- subset(df.bgSNPs_w_phenotypes, df.bgSNPs_w_phenotypes[,trait] == "yes")
  
  v.bQTLs.part <- colSums(ifelse(df.bQTLs.trait[,v.partitions] == "yes", 1, 0))
  v.bgSNPs.part <- colSums(ifelse(df.bgSNPs.trait[,v.partitions] == "yes", 1, 0))
  
  sampleSize <- nrow(df.bQTLs.trait)
  popSize <- nrow(df.bgSNPs.trait)
  
  for(part in v.partitions){
    
    hitInSample <- v.bQTLs.part[part]
    hitInPop <- v.bgSNPs.part[part]
    failInPop <- popSize - hitInPop
    fc <- ((hitInSample / sampleSize) / (hitInPop / popSize))
    p.val <- phyper(hitInSample, hitInPop, failInPop, sampleSize, lower.tail = FALSE)

    d.gsea$trait[idx] <- trait
    d.gsea$partition[idx] <- part
    d.gsea$p.val[idx] <- p.val
    d.gsea$number[idx] <- hitInSample
    d.gsea$percent[idx] <- round(hitInSample / sampleSize * 100, 1)
    d.gsea$foldchange[idx] = fc
    
    idx <- idx + 1
    
  }
  
}
  
write.csv(d.gsea, "2000bp_dist_phenotypic_gwas_6143_ASBs_317094_bgSNPs_B73_partition_enrichment.csv",  row.names = FALSE, quote = FALSE)

d.gsea <- perform_gwas(df.ASBs, df.bgSNPs, B73.only = T)
write.csv(d.gsea, "2000bp_dist_phenotypic_gwas_6143_ASBs_317094_bgSNPs_B73_postfreq_only.csv",  row.names = FALSE, quote = FALSE)



message("Genomic feature profiling of ASBs (Methylation, Motifs and DNAseI, Enhancers) ")





message("------------ Methylation ---------")

# To identify ASBs which overlapped with significant variation in either CG, CHG, or CHH
# methylation between B73 and Mo17, we first, per ASB, assigned averaged methylation
# levels of Mo17 and B73 methylation levels (separately for the CG, CHG, or CHH
#  methylation datasets) within a given window of +/- 40 bp around the ASB position.
# Differentially methylated alleles were defined as described previously 13 . Accordingly, we
# identified ASBs as overlapping with differentially methylated regions if in the B73 or Mo17
# methylation datasets, the methylation level of one allele would be >= 70% while the level
# of the corresponding allele was <= 10%.


# Methylation levels for CG, CHG, and CHH for B73 and Mo17 were extracted from Regulski et al. 2013
# Methylation frequency versus distance (up to +/- 2 kbp) around
# each ASB were averaged over 20bp bins, and visualized by regions bound by BZR1 with
# either high or low affinity levels depending on the inbred line. For B73, high and low
# affinity bound regions were defined by a post frequency of >= 0.85 or <= 0.15, respectively
# and oppositely for Mo17 by a post frequency <= 0.15 and >= 0.85), respectively.


source("utils/methylation_revision.R")
v.parents = c("B73", "Mo17", "B73", "Mo17", "B73", "Mo17")
v.variants = c("CG", "CG", "CHH", "CHH", "CHG", "CHG")

folder_tmp_methylation <- "tmp_revision/methylation_hybrid"

# hybrid
v.filenames = c("data/revision/methylation/BxM.CG_sorted.diploid.bw", 
                "data/revision/methylation/BxM.CG_sorted.diploid.bw",
                "data/revision/methylation/BxM.CHH_sorted.diploid.bw",
                "data/revision/methylation/BxM.CHH_sorted.diploid.bw",
                "data/revision/methylation/BxM.CHG_sorted.diploid.bw",
                "data/revision/methylation/BxM.CHG_sorted.diploid.bw")

# parents
# v.filenames = c("data/revision/methylation/B73_methylation_CG.bw", 
#                 "data/revision/methylation/Mo17_methylation_CG.bw",
#                 "data/revision/methylation/B73_methylation_CHH.bw",
#                 "data/revision/methylation/Mo17_methylation_CHH.bw",
#                 "data/revision/methylation/B73_methylation_CHG.bw",
#                 "data/revision/methylation/Mo17_methylation_CHG.bw")
v.formats = c("bigWig", "bigWig", "bigWig", "bigWig", "bigWig", "bigWig")
v.has_ranges = c(FALSE, FALSE, FALSE, FALSE, FALSE, FALSE)
th.bp_offset = 20


df.ASBs <- readRDS(paste(folder_tmp, "6143_ASBs_w_genomic_location.rds", sep = "/"))
df.bgSNPs <- readRDS(paste(folder_tmp, "317094_bgSNPs_w_genomic_location.rds", sep = "/"))

# PERFORM METHYLATION OF ALL 
methylation_occupancy(df.ASBs, 
                     df.bgSNPs,
                     v.filenames,
                     v.formats,
                     v.parents,
                     v.variants,
                     v.has_ranges,
                     th.bp_offset,
                     n.chromosomes = 10,
                     n.cpus = 1,
                     folder = folder_tmp_methylation)


res <- load_methylation(df.ASBs, 
                        df.bgSNPs,
                        v.parents,
                        v.variants,
                        folder = folder_tmp_methylation)

df.ASBs <- res$df.ASBs
df.bgSNPs <- res$df.bgSNPs

write.csv(df.ASBs, "6143_ASBs_w_genomic_location_and_methylation.csv", quote = F, row.names = F)
write.csv(df.bgSNPs, "317094_bgSNPs_w_genomic_location_and_methylation.csv", quote = F, row.names = F)

dev.off()
k = 2
title = paste("allelic biases vs methylation levels (averaged per ASB using ",  th.bp_offset, "bp window)", sep = "")
par(mfrow=c(3,2))
scatter_plot(df.ASBs$POSTfreq, df.ASBs$`CG-B73`, xlab = "allelic bias", ylab = "methylation level (B73 CpG)")
scatter_plot(df.ASBs$POSTfreq, df.ASBs$`CG-Mo17`, xlab = "allelic bias", ylab = "methylation level (MO17 CpG)")
scatter_plot(df.ASBs$POSTfreq, df.ASBs$`CHG-B73`, xlab = "allelic bias", ylab = "methylation level (B73 CHG)")
scatter_plot(df.ASBs$POSTfreq, df.ASBs$`CHG-Mo17`, xlab = "allelic bias", ylab = "methylation level (MO17 CHG)")
scatter_plot(df.ASBs$POSTfreq, df.ASBs$`CHH-B73`, xlab = "allelic bias", ylab = "methylation level (B73 CHH)")
scatter_plot(df.ASBs$POSTfreq, df.ASBs$`CHH-Mo17`, xlab = "allelic bias", ylab = "methylation level (MO17 CHH)")
mtext(title, side = 3, line = -2, outer = TRUE, cex = 0.8)



# repeat with displaying only methylation explained ASBs
dev.off()
title = paste("allelic biases vs methylation levels (averaged per ASB using ",  th.bp_offset, "bp window)", sep = "")
par(mfrow=c(2,2))
idx = which(df.ASBs$`CG-Mo17` <= 0.3 & df.ASBs$`CG-B73` >= 0.7)
scatter_plot(df.ASBs$POSTfreq[idx], df.ASBs$`CG-B73`[idx], xlab = "allelic bias", ylab = "methylation level (B73 CpG)")

idx = which(df.ASBs$`CG-Mo17` >= 0.7 & df.ASBs$`CG-B73` <= 0.3)
scatter_plot(df.ASBs$POSTfreq[idx], df.ASBs$`CG-Mo17`[idx], xlab = "allelic bias", ylab = "methylation level (MO17 CpG)")

idx = which(df.ASBs$`CHG-Mo17` <= 0.3 & df.ASBs$`CHG-B73` >= 0.7)
scatter_plot(df.ASBs$POSTfreq[idx], df.ASBs$`CHG-B73`[idx], xlab = "allelic bias", ylab = "methylation level (B73 CHG)")

idx = which(df.ASBs$`CHG-Mo17` >= 0.7 & df.ASBs$`CHG-B73` <= 0.3)
scatter_plot(df.ASBs$POSTfreq[idx], df.ASBs$`CHG-Mo17`[idx], xlab = "allelic bias", ylab = "methylation level (MO17 CHG)")

mtext(title, side = 3, line = -2, outer = TRUE, cex = 0.8)


### 

dev.off()
k = 2
title = paste("allelic biases vs differential methylation levels (averaged per ASB using ",  th.bp_offset, "bp window)", sep = "")
par(mfrow=c(3,1))

idx1 = (df.ASBs$`CG-Mo17` <= 0.1 & df.ASBs$`CG-B73` >= 0.7)
idx2 = (df.ASBs$`CG-Mo17` >= 0.7 & df.ASBs$`CG-B73` <= 0.1)
idx = which(idx1 | idx2)
col_vals = rep("black", nrow(df.ASBs)) 
col_vals[idx] = "red"

scatter_plot(df.ASBs$POSTfreq, df.ASBs$`CG-Mo17` - df.ASBs$`CG-B73`, 
             xlab = "allelic bias", ylab = "methylation level difference (CpG)", 
             x_limit = c(0,1), y_limit = c(-1,1), 
             col_vals = col_vals)

### 
idx1 = (df.ASBs$`CHG-Mo17` <= 0.1 & df.ASBs$`CHG-B73` >= 0.7)
idx2 = (df.ASBs$`CHG-Mo17` >= 0.7 & df.ASBs$`CHG-B73` <= 0.1)
idx = which(idx1 | idx2)
col_vals = rep("black", nrow(df.ASBs)) 
col_vals[idx] = "red"

scatter_plot(df.ASBs$POSTfreq, 
             df.ASBs$`CHG-Mo17` - df.ASBs$`CHG-B73`,
             xlab = "allelic bias", ylab = "methylation level difference (CHG)", 
             x_limit = c(0,1), y_limit = c(-1,1), 
             col_vals = col_vals)



### 
idx1 = (df.ASBs$`CHH-Mo17` <= 0.1 & df.ASBs$`CHH-B73` >= 0.7)
idx2 = (df.ASBs$`CHH-Mo17` >= 0.7 & df.ASBs$`CHH-B73` <= 0.1)
idx = which(idx1 | idx2)
col_vals = rep("black", nrow(df.ASBs)) 
col_vals[idx] = "red"

scatter_plot(df.ASBs$POSTfreq, 
             df.ASBs$`CHH-Mo17` - df.ASBs$`CHH-B73`,
             xlab = "allelic bias", ylab = "methylation level difference (CHH)", 
             x_limit = c(0,1), y_limit = c(-1,1), 
             col_vals = col_vals)

mtext(title, side = 3, line = -2, outer = TRUE, cex = 0.8)



####

methylation_occupancy_distance_to_asb(df.ASBs, 
                                      v.filenames,
                                      v.formats,
                                      v.parents,
                                      v.variants,
                                      v.has_ranges,
                                      n.bin_width = 10,
                                      th.distance_to_ASB = 2000,
                                      th.high_affinity = 0.85,
                                      n.chromosomes = 10,
                                      group = "all",
                                      n.cpus = 1,
                                      folder = folder_tmp_methylation)


plot_methylation_occupancy_distance_to_asb(v.variant_sets = c("CG", "CHG", "CHH"),
                                           th.high_affinity = 0.85,
                                           folder = folder_tmp_methylation)

# 
# methylation_occupancy(l.bQTL_gene_partitioning, 
#                       th.distance_to_ASB = 2000,
#                       n.chromosomes = 10,
#                       n.cpus = 4,
#                       n.binWidth = 20,
#                       b.val = TRUE,
#                       v.species = c("Mo17", "B73"),
#                       v.groups = c("all", "genic", "non_genic"),
#                       v.species = c("Mo17", "Mo17" , "B73", "B73"),
#                       v.stringency <- c("< 0.5 | > 0.5", "< 0.15 | > 0.85", "< 0.5 | > 0.5", "< 0.15 | > 0.85"),
#                       v.filenames = c(paste(folder_data,"methylation/B73_CHG.bw", sep = "/"),
#                                       paste(folder_data,"methylation/B73_CHH.bw", sep = "/"),
#                                       paste(folder_data,"methylation/B73_CpG.bw", sep = "/"),
#                                       paste(folder_data,"methylation/MO17_CHG.bw", sep = "/"),
#                                       paste(folder_data,"methylation/MO17_CHH.bw", sep = "/"),
#                                       paste(folder_data,"methylation/MO17_CpG.bw", sep = "/")),
#                       v.datasets = c("B73_CHG", "B73_CHH", "B73_CpG", "MO17_CHG", "MO17_CHH", "MO17_CpG"),
#                       v.formats = c("bigWig", "bigWig", "bigWig", "bigWig", "bigWig", "bigWig"),
#                       folder_tmp = folder_tmp,
#                       folder_output = folder_output,
#                       do.plot = F)
# 
# 
# # rename methylation, open chromatin?
# l.SNPs_w_chromatin <- RDS(paste(folder_tmp, "l.SNPs_w_chromatin_ASBs_and_bgSNPs.rds", sep = "/"))
# 
# # Fig. 4 l) Correlation of CG methylation with allele-specific BZR1 binding. Average
# # CG methylation of the B73 - Mo17 allele of the 40bp surrounding each ASB are plotted against
# # the allelic bias (expressed in percentage of B73 read counts). Differential CG methylation is
# # indicated by red dots.
# differential_methylation_vs_allelic_bias(l.SNPs=l.bQTL_gene_partitioning,
#                                          th.bp_offset = 20,
#                                          th.distance_to_ASB = 2000,
#                                          n.chromosomes = 10,
#                                          n.cpus = 1,
#                                          b.val = TRUE,
#                                          v.species = c("Mo17", "B73"),
#                                          degrees = 15,
#                                          th.distance_to_ASB = 5000,
#                                          th.padding = 50,
#                                          width = 6,
#                                          height = 4,
#                                          group = "genic_and_nongenic",
#                                          v.filenames = c("data/dnase/GSE94291_DNase_ist.bedGraph", 
#                                                          "data/methylation/B73_CpG.bw", 
#                                                          "data/methylation/MO17_CpG.bw",
#                                                          "data/methylation/B73_CHG.bw",
#                                                          "data/methylation/MO17_CHG.bw",
#                                                          "data/methylation/B73_CHH.bw",
#                                                          "data/methylation/MO17_CHH.bw"),
#                                          v.datasets = c("DNase", "B73_CpG",  "MO17_CpG", "B73_CHG", "MO17_CHG", "B73_CHH", "MO17_CHH"),
#                                          v.formats = c("bedGraph", "bigWig", "bigWig", "bigWig", "bigWig", "bigWig", "bigWig"),
#                                          v.has_ranges = c("yes",  "no", "no",  "no", "no",  "no", "no")
# )
# 
# 
# # TODO: get absolute and differential figure ... 
# 
# 
# l.SNPs_w_chromatin <- readRDS(paste(folder_tmp, "l.SNPs_w_chromatin_ASBs_and_bgSNPs.rds", sep = "/"))
# 
# 
# df.ASBs
# 
# 
# 
# l.SNPs=l.bQTL_gene_partitioning
# th.bp_offset = 20
# th.distance_to_ASB = 2000
# n.chromosomes = 10
# n.cpus = 1
# b.val = TRUE
# v.species = c("Mo17", "B73")
# degrees = 15
# th.distance_to_ASB = 5000
# th.padding = 50
# width = 6
# height = 4
# group = "genic_and_nongenic"
# v.filenames = c("data/dnase/GSE94291_DNase_ist.bedGraph", 
#                 "data/methylation/B73_CpG.bw", 
#                 "data/methylation/MO17_CpG.bw",
#                 "data/methylation/B73_CHG.bw",
#                 "data/methylation/MO17_CHG.bw",
#                 "data/methylation/B73_CHH.bw",
#                 "data/methylation/MO17_CHH.bw")
# v.datasets = c("DNase", "B73_CpG",  "MO17_CpG", "B73_CHG", "MO17_CHG", "B73_CHH", "MO17_CHH")
# v.formats = c("bedGraph", "bigWig", "bigWig", "bigWig", "bigWig", "bigWig", "bigWig")
# v.has_ranges = c("yes",  "no", "no",  "no", "no",  "no", "no")
# 


# Fig. 2. Variation in DNA sequence and methylation underlie differential BZR1 binding. a)
# Schematics of HASCh-seq approach and possible causes for allele-specific binding events.
# Chromatin-IP is performed in F1 hybrid plants. From top to bottom, three possible scenarios for
# TF binding to the parental genomes (green and blue) are depicted, where binding strength is
# represented by the line width of the black arrows: when no alterations in motif and chromatin
# structure occurs, binding is expected to be equal. Lower binding is expected If the binding motif
# is altered by a SNP or epigenetic features like DNA methylation for one of the alleles. b) An
# example of allele-specific binding of ZmBZR1 to target gene Zm00001d031434. BZR1 bound
# target reads that uniquely map to either B73 (green) or Mo17 (blue) are shown. c) Allelic and
# spatial distribution of allele-specific BZR1 binding sites with significant bias towards B73 (green)
# or Mo17 (blue) along the maize B73 reference genome. Allelic bias is expressed as percentage
# of B73 read counts. Chromosome borders and length are depicted by dashed lines and arrows at
# the bottom, respectively. Centromere locations are indicated by orange rectangles. A red box on
# 26chromosome 1 marks the ASB upstream of Zm00001d031434 displayed in Fig. 2b. d) Genomic
# distribution of ASBs. 3297 ASBs were classified according to their location relative to genes. In
# case of two genes in the proximity of an ASB, the priority given was exon>intron>UTR>1 kb
# upstream>1 kb downstream>1-5 kb upstream. e) Frequency of BRRE (CGTG[C/T]G), G-Box
# (CACGTG), and the control motif GTACGG (SBP-box 12 ) around ASBs of the alleles with higher
# BZR1 binding. f) Fraction of ASBs overlapping with motifs, for which the allele with canonical
# BRRE, G-Box or a control motif GCCGCC (GCC-box 12 ), showed higher ZmBZR1 affinity. Both
# BR-related BRRE and G-box motifs, but not the control motif, diverge significantly (p<0.001,
# Fisherâ€™s exact test) from the expected 50% random distribution. g) Distribution of lead SNPs of
# ASBs within motifs. Among ASBs which overlapped with BRRE motif, SNPs were enriched in the
# core CG bases. h) ASBs affecting BZR1 motifs and/or overlapping with differentially methylated
# (CpG, CHG or CHH) sites between B73 and Mo17. i)-k) Average CpG (h), CHG (i) and CHH(j)
# methylation frequency in B73 (green) and Mo17 (blue) over ASB loci with a least 85% binding
# bias towards B73 or Mo17. High affinity ( ___ ) and low affinity ( â€¦. ) alleles are considered separately
# for each genotype. l) Correlation of CG methylation with allele-specific BZR1 binding. Average
# CG methylation of the B73 - Mo17 allele of the 40bp surrounding each ASB are plotted against
# the allelic bias (expressed in percentage of B73 read counts). Differential CG methylation is
# indicated by red dots.







message("------------ Motifs ---------")
# To identify ASBs which may be explained by motif variation (Fig. 2h), we extracted the
# +/- 5 bp of the high affinity BZR1 bound allele surrounding ASBs. Using R we scanned
# those 11 bp fragments for canonical BRRE (CGTG[T/C]G, C[G/A]CACG), allowing a
# single base pair mismatch outside the core motif (CGTG), or G-box (CACGTG) motifs
# and determined ASBs where the SNP changed a BRRE or G-box motif into an altered
# (non BRRE or G-box) motif. 

df.ASBs <- readRDS(paste(folder_tmp, "6143_ASBs_w_genomic_location.rds", sep = "/"))

source("utils/motifs_revision.R")
res <- predefined_motif_analysis(df.ASBs, v.motifs, s.env_bps = 5, n.chromosomes = 10,
                                 path.genomic_sequences = "data/revision/ref_B73Mo17.fasta")

write.csv(res$df.motif_asbs, "motif_directionality_6143_ASBs.csv")
write.csv(res$df.bQTLs, "6143_ASBs_w_motif_annotation.csv", quote = F, row.names = F)
write.csv(as.data.frame(res$m.motif_variation), "motif_base_6143_ASBs_variation.csv")
print(res$m.motif_variation)

res$m.motif_variation <- round(res$m.motif_variation / rowSums(res$m.motif_variation) * 100, 2) 
print(res$m.motif_variation)

write.csv(as.data.frame(res$m.motif_variation), "motif_base_6143_ASBs_variation_percentage.csv")


# compare motifs between ASBs and bQTLs (in peaks before linkage and blacklisting)

df.bQTLs <- readRDS(paste(folder_tmp, "33267_bQTLs_in_peaks.rds", sep = "/"))

res <- predefined_motif_analysis(df.bQTLs, v.motifs, s.env_bps = 5, n.chromosomes = 10,
                                 path.genomic_sequences = "data/revision/ref_B73Mo17.fasta")

write.csv(res$df.motif_asbs, "motif_directionality_33267_bQTLs_in_peaks.csv")
write.csv(res$df.bQTLs, "33267_bQTLs_in_peaks_w_motif_annotation.csv", quote = F, row.names = F)
write.csv(as.data.frame(res$m.motif_variation), "motif_base_33267_bQTLs_in_peaks_variation.csv")
print(res$m.motif_variation)

res$m.motif_variation <- round(res$m.motif_variation / rowSums(res$m.motif_variation) * 100, 2) 
print(res$m.motif_variation)

write.csv(as.data.frame(res$m.motif_variation), "motif_base_33267_bQTLs_in_peaks_variation_percentage.csv")

# check comparison 


# 
# 
# if(!b.load_motif_analysis){
#   source("motifs/predefined_motif_analysis.R")
#   
#   
#   
#   predefined_motif_analysis(l.bQTL_gene_partitioning, df.peaks, motifs, v.motif_offset)
#   
#   l.motif_analysis <- readRDS(paste(folder_tmp, "l.motif_analysis.rds", sep = "/"))
#   l.nucleotideInPeaks <- readRDS(paste(folder_tmp, "l.nucleotideInPeaks.rds", sep = "/"))
#   
#   
#   source("motifs/predefined_motifs_in_peaks.R")
#   predefined_motifs_in_peaks(l.motif_analysis, l.nucleotideInPeaks)
#   
#   l.motif_analysis.postprocessed = readRDS(paste(folder_tmp, "l.motif_analysis.postprocessed.rds", sep = "/"))
#   
#   
#   source("motifs/directionality_and_snp_distribution.R")
#   directionality_and_snp_distribution(l.motif_analysis.postprocessed)
#   
#   df.motif_nucleotidePositionPartitions <- readRDS(paste(folder_tmp, "df.motif_nucleotidePositionPartitions.rds", sep = "/"))
#   
#   
#   source("motifs/motif_snp_position_significance.R")
#   motif_snp_position_significance(l.bQTL_gene_partitioning, motifs, v.motif_offset, df.peaks, genome, l.genome.mutant, n.chromosomes)
#   
#   
#   source("motifs/evaluating_BZR1_core_motif.R")
#   evaluating_BZR1_core_motif(df.ASBs, core_motif = "CGTG", s.motif_offset = 5, n.chromosomes = 10)
#   
# }else{
#   l.motif_analysis <- readRDS(paste(folder_tmp, "l.motif_analysis.rds", sep = "/"))
#   l.nucleotideInPeaks <- readRDS(paste(folder_tmp, "l.nucleotideInPeaks.rds", sep = "/"))
#   l.motif_analysis.postprocessed = readRDS(paste(folder_tmp, "l.motif_analysis.postprocessed.rds", sep = "/"))
#   df.motif_nucleotidePositionPartitions <- readRDS(paste(folder_tmp, "df.motif_nucleotidePositionPartitions.rds", sep = "/"))
#   df.motifPositionAnalysis <- readRDS(paste(folder_tmp, "df.motifPositionAnalysis.rds", sep = "/"))
#   df.ASBs.core_motifs <- readRDS(paste(folder_tmp, "df.ASBs.core_motifs.rds", sep = "/"))
# }
# 
# if(b.write_paper){
#   write.table(l.motif_analysis[[1]], paste(folder_output, "/motifs/S0_1.txt", sep = ""),  row.names = FALSE, quote = FALSE, sep ="\t")
#   
#   print(table(l.motif_analysis.postprocessed[[1]]$direction))
#   write.table(l.motif_analysis.postprocessed[[1]], paste(folder_output, "/motifs/S0_2.txt", sep = ""),  row.names = FALSE, quote = FALSE, sep ="\t")
#   write.table(df.motif_directionality, paste(folder_output, "/motifs/S0_3.txt", sep = ""),  row.names = FALSE, quote = FALSE, sep ="\t")
#   write.table(df.motif_nucleotidePositionPartitions, paste(folder_output, "/motifs/S0_4.txt", sep = ""),  row.names = FALSE, quote = FALSE, sep ="\t")
#   write.table(df.motifPositionAnalysis, paste(folder_output, "/motifs/S0_5.txt", sep = ""),  row.names = FALSE, quote = FALSE, sep ="\t")
#   write.table(df.ASBs.core_motifs, paste(folder_output, "/SX3.txt", sep = ""), quote = FALSE, row.names = FALSE, sep ="\t")
#   
# }
# 


message("------------ Enhancer ---------")
source("utils/enhancer.R")


df.ASBs <- readRDS(paste(folder_tmp, "6143_ASBs_w_genomic_location.rds", sep = "/"))      
df.bgSNPs <- readRDS(paste(folder_tmp, "317094_bgSNPs_w_genomic_location.rds", sep = "/"))

res <- snp_to_enhancer_distance(df.ASBs,
                               df.bgSNPs,
                               path.enhancer,
                               n.binWidth = 10,
                               n.maxDist = 10000)


write.csv(res$df.ASBs_w_enhancer_annotation, "2730_non_genic_ASBs_w_genomic_location_and_enhancer.csv", quote = F, row.names = F)
write.csv(res$df.bgSNPs_w_enhancer_annotation, "136500_non_genic_bgSNPs_w_genomic_location_and_enhancer.csv", quote = F, row.names = F)

df.ASBs_enhancer <- subset(res$df.ASBs_w_enhancer_annotation, res$df.ASBs_w_enhancer_annotation$non_genic == "yes")
table(df.ASBs_enhancer$in_enhancer_region)

df.bgSNPs_enhancer <- subset(res$df.bgSNPs_w_enhancer_annotation, res$df.bgSNPs_w_enhancer_annotation$non_genic == "yes")
table(df.bgSNPs_enhancer$in_enhancer_region)

(table(df.ASBs_enhancer$in_enhancer_region)[2]/nrow(df.ASBs_enhancer))/(table(df.bgSNPs_enhancer$in_enhancer_region)[2]/nrow(df.bgSNPs_enhancer))



# # TODO: get both ... 
# 
# message("--- Expression --- ")
# 
# # TODO: make with correct gene (double) annotations 
# 
# length(unique(df.ASBs$gene.ID))
# 
# df.expression <- read.table(path.expression, header = FALSE, sep = "\t", stringsAsFactors = FALSE, fill = TRUE)


