rm(list=ls()) # clear workspace 

folder_output = "output_test" 
folder_tmp = "tmp_test"
folder_data = "data"

# TODO: wichtig - code figures etc. bezug auf current state article !!!! (get pdf)

# TODO: define an input SNP dataset 

b.write_paper = FALSE
# TODO: make a readme table for mapping tables to ids 


# TODO: create folder 
# TODO: find ZmvsAth_1_to_1_corrected_12084.csv

# TODO: which phenotypes dataa to be used
# TODO: enhancer data?


source("config.R")

ifelse(!dir.exists(file.path(folder_output)), dir.create(file.path(folder_output)), FALSE)
ifelse(!dir.exists(file.path(folder_tmp)), dir.create(file.path(folder_tmp)), FALSE)


source("utils.R")
install_and_load_libraries()


# define input datasets 
message("load and preprocess basic datasets...")
source("datasets.R")


# BACKGROUND
# 
# Variation in transcriptional regulation is a major cause of phenotypic diversity.
# Genomewide association studies (GWAS) have shown that most functional variants reside in 
# noncoding regions, where they potentially affect transcription factor (TF) binding and
# chromatin accessibility to alter gene expression. Pinpointing such regulatory variations,
# however, remains challenging2. Here, we developed a hybrid allele-specific chromatin
# binding sequencing (HASCh-seq) approach and identified variations in target binding of
# the brassinosteroid (BR) responsive transcription factor ZmBZR1 in maize. Chromatin
# immunoprecipitation followed by sequencing (ChIP-seq) in B73xMo17 F1s identified
# thousands of target genes of ZmBZR1. Allele-specific ZmBZR1 binding (ASB) was
# observed for about 14.3% of target genes. It correlated with over 550 loci containing
# sequence variation in BZR1-binding motifs and over 340 loci with haplotype-specific DNA
# methylation, linking genetic and epigenetic variations to ZmBZR1 occupancy.
# Comparison with GWAS data linked hundreds of ASB loci to important yield, growth and
# disease-related traits. Our study provides a robust method for analyzing genome-wide
# variations of transcription factor occupancy and identified genetic and epigenetic
# variations of the BR response transcription network in maize. 


message("--- Hasch-seq data analysis ---")

# Mitigating potential biases are critical steps in order to accurately analyze allele specific
# binding. For HASCh-seq, we first tried to address mapping bias using an enhanced
# reference approach previously described. 

# First, we generate a pseudo Mo17 reference genome by integrating all homozygous B73/Mo17 HapMap 3.2.1 SNPs into the B73
# AGPv4 genome (R v.3.3.2), 
if(b.firstRun){
  source("utils/mutant_genome.R")
  create_mutant_genome_from_vcf()
}else{
  df.snp_positions <- read.table(paste(folder_data,"df.snp_positions.csv", sep = "/"),  sep ="\t")
  l.genome.mutant <- readRDS(paste(folder_tmp,"genome_mutant.rds", sep = "/"))
}
message(nrow(df.snp_positions), " SNPs in the pseudo Mo17 reference genome (generated by integrating all homozygous B73/Mo17 HapMap 3.2.1 SNPs into the B73 AGPv4 genome) ")


# -- EXTERNAL PROCEDURE ---
# Then, we mapped quality filtered paired-end reads against the B73
# and pseudo Mo17 reference genomes with bwa-mem (v. 0.7.16a) using default
# parameters. Next, we merged the corresponding aligned reads, retaining uniquely or
# higher quality mapping reads to either B73 or Mo17, or a random read in case of equal
# mapping quality using the merge-sam script7. Finally, we filtered for uniquely mapped
# reads (q13), and removed potential PCR duplicates (retaining a random read, instead
#  higher mapping quality to avoid B73 bias). Overlapping mate reads were trimmed to avoid
# duplicated counting using the clip_overlap script. Bias corrected mapping files
# were used to determine both BZR1 binding peaks using the GEM pipeline described
# above, as well as ASBs. To determine high confidence peaks, samples were analyzed
# individually and peaks reproducible in all 6 replicates, using a 400 bp window (+/- 200 bp around each peak) were retained. 

# To identify regions with differential TF binding, we first merged bias-corrected bam files
# for IP or Input replicates and determined allele frequencies for all homozygotes HapMap
# 3.2.1. B73/Mo17 SNPs (6,685,773) determined using ASEReadCounter (GATK 3.6). 

# ZmBZR1-YFP ASBs were determined using custom R (v. 3.3.2) scripts. In order to
# accurately access allele frequencies of all homozygous SNPs, we set a minimum read 
# coverage cutoff of ≥ 50 reads (n=162,398). Next, to reduced potential artifacts due to false
# positive B73/Mo17 SNPs, base calling errors in the B73 AGPv4 reference genome and
# sequencing errors, we removed SNPs with a bias ≥ 90% to either parent, if both the B73
# and Mo17 allele could not be verified in the input data (n=160,755 SNPs). To further
# reduce potential background or artificial bias effects, we blacklisted genomic regions (chr.
# 3: 186,240,476- 210,080,072; chr10: 2,965,337- 4124958) with a systematic bias in the
# input data. Resulting in 151,073 SNPs.
source("utils/snp_preprocessing.R")
message(nrow(postTotal), " SNPs") 

 

# Of the remaining 151,073 SNPs, we determined significant variation of median allele
# frequency of 0.515 using a binomial test with a p-value cutoff of ≤ 0.001 adjusted for
# multiple testing using Bonferroni correction (n=19,601 SNPs)
l.postTotal <- vector(mode = "list", length = 2)
l.postTotal[[1]] <- subset(postTotal, postTotal$`p-value (corrected)` < th.p.bQTL) # significant bQTL
l.postTotal[[1]] <- subset(l.postTotal[[1]], l.postTotal[[1]]$POSTfreq < 1 & l.postTotal[[1]]$POSTfreq > 0)
if(b.save_intermediate_results){
  write.table(l.postTotal[[1]], paste(folder_output, "/ASBs.txt", sep = ""),  row.names = FALSE, quote = FALSE, sep ="\t")
}
message("bQTL (significance p < ", th.p.bQTL, ", ", nrow(l.postTotal[[1]]), " SNPs)")
hist(l.postTotal[[1]]$POSTfreq, breaks = 100, main = "Allelic bias of bQTL", xlab = "allelic bias")





message("loading binding peak datasets...")
# To focus on ASBs (Allele-specific ZmBZR1 binding) with potential biological relevance, we further restricted ASBs to those
# located in highconfidence BZR1-binding peaks reproducible in all 6 biological replicates

if(!b.load_filtered_binding_peaks){
  source("utils/binding_peaks.R") # TODO: DEFINE AND EXPLAIN THE FUNCTION - how to get to final peaks
}else{
  df.peaks.final <- readRDS(paste(folder_tmp,"df.peaks.final.rds", sep ="/")) # TABLE S5 # TODO: what is the table definition
}
df.peaks <- df.peaks.final
df.peaks["start"] <- df.peaks$posPeak - s.half_window_size
df.peaks["end"] <- df.peaks$posPeak + s.half_window_size
message(nrow(df.peaks), " high-confidence BZR1 binding peaks")

if(b.peak_analysis){ # Note: the list type is actually obsolete (previously estimated for both ASB and population)
  
  message("Estimate significant bQTLs in binding peak ranges...")
  
  l.selection <- vector(mode = "list", length = 2)
  l.sigSnps_in_peaks <- vector(mode = "list", length = 2)
  
  df.non_ASB_peaks <- c()
  
  for(s in 1:1){
    
    n.total <- 0
    postTotal.significant <- l.postTotal[[s]]
    
    strt<-Sys.time() 
    cl<-makeCluster(min(n.chromosomes, n.cpus))
    registerDoParallel(cl)
    l.res <- foreach(i = 1:n.chromosomes, .packages = c("seqinr", "Biostrings", "VariantAnnotation")) %dopar% {     
      
      sigSnps_in_peaks <- 0
      df.selection <- c()
      
      df.snp.pos <- readRDS(paste(folder_tmp,"/df.snp.pos_",i, ".rds", sep = ""))
      
      # remove heterozygote snps
      if(FALSE){
        idx.snps <- which(lapply(df.snp.pos@elementMetadata$ALT, length) == 1)  
      }else{
        vec.snp.bases <- CharacterList(df.snp.pos@elementMetadata$ALT)
        vec.snp.bases = unstrsplit(vec.snp.bases, sep = ",")
        vec.snp.bases <- gsub("\\,.*", "", vec.snp.bases)
        df.snp.pos@elementMetadata$ALT <- vec.snp.bases
      }
      
      # A - generic SNPs
      if(FALSE){
        vec.snp.pos <- as.numeric(df.snp.pos@ranges@start[idx.snps])  
        vec.snp.bases <- as.character(unlist(df.snp.pos@elementMetadata$ALT[idx.snps]))
      }else{
        vec.snp.pos <- as.numeric(df.snp.pos@ranges@start)  
        vec.snp.bases <- as.character(unlist(df.snp.pos@elementMetadata$ALT))
      }
      
      vec.snp.bases <- ifelse(vec.snp.bases == "<DEL>", "N",vec.snp.bases) # both represented separately
      vec.snp.bases <- ifelse(vec.snp.bases == "<INS>", "N",vec.snp.bases)
      
      n.total <- n.total + length(vec.snp.pos)
      
      # remove empty strings # 
      idx.snp.exceptions <- which(vec.snp.bases == "")
      
      if(length(idx.snp.exceptions) > 0){
        vec.snp.pos <- vec.snp.pos[-idx.snp.exceptions]
        vec.snp.bases <- vec.snp.bases[-idx.snp.exceptions]
      }
      
      genome.reference <- DNAString(genome[[i]])
      genome.mutant    <- replaceLetterAt(genome.reference, vec.snp.pos, vec.snp.bases) # vergleich vor austausch
      
      # binding peaks 
      if(s == 1){
        tf_target_bind.sset <- subset(df.peaks, df.peaks$seqnames == i)
      }else if(s == 2){
        tf_target_bind.sset <- subset(df.non_ASB_peaks, df.non_ASB_peaks$seqnames == i)
      }
      
      peak <- tf_target_bind.sset
      
      #   for(p in 1:nrow(peak)){ # double check for all
      #     idx.snps <- which(vec.snp.pos >= peak$start[p] & vec.snp.pos <= peak$end[p])
      #     snp_in_peaks <- snp_in_peaks + length(idx.snps)  
      #   }
      
      postTotal.significant.i <- subset(postTotal.significant ,postTotal.significant$contig == i)
      
      idx.non_asb_peaks <- numeric()
      
      pb <- txtProgressBar(min = 0, max = nrow(peak), style = 3)
      
      for(p in 1:nrow(peak)){ # double check for all
        
        setTxtProgressBar(pb, p)
        
        idx.snps <- which(postTotal.significant.i$position >= peak$start[p] & postTotal.significant.i$position <= peak$end[p])
        
        if(length(idx.snps) > 0){
          df.selection <- rbind(df.selection ,postTotal.significant.i[idx.snps,])
          sigSnps_in_peaks <- sigSnps_in_peaks + length(idx.snps)  
        }else{
          idx.non_asb_peaks <- c(idx.non_asb_peaks, p)
        }
        
      }
      
      close(pb)
      
      l.res <- list(df.selection = df.selection, sigSnps_in_peaks = sigSnps_in_peaks, df.non_ASB_peaks = peak[idx.non_asb_peaks,])
      
    }
    stopCluster(cl)
    print(Sys.time()-strt)
    
    l.selection[[s]] <- (l.res[[1]]$df.selection)
    l.sigSnps_in_peaks[[s]] <- (l.res[[1]]$sigSnps_in_peak)
    
    df.non_ASB_peaks <- l.res[[1]]$df.non_ASB_peaks
    
    for(i in 2:n.chromosomes){
      l.selection[[s]] <- rbind(l.selection[[s]], (l.res[[i]]$df.selection))
      l.sigSnps_in_peaks[[s]] <- l.sigSnps_in_peaks[[s]] + (l.res[[i]]$sigSnps_in_peak)
      df.non_ASB_peaks <- rbind(df.non_ASB_peaks, l.res[[i]]$df.non_ASB_peaks)
    }
  }
  
  message("...finished")
  message("ASBs" , nrow(l.selection[[1]]))
  
  if(b.save_intermediate_results){
    write.table(l.selection[[1]], paste(folder_output, "/bQTL_after_input_blacklisting_pvalue_filter_in_peaks.txt", sep = ""), quote = FALSE, row.names = FALSE, sep ="\t")
  }
  
}
n_snps_peak_filtered = nrow(l.selection[[1]])
message(n_snps_peak_filtered, " SNPs located within high-confidence BZR1 binding peaks")



# SNPs in close proximity with causative polymorphisms will show biased allele frequency due to linkage
# disequilibrium (LD). To address this, we identified SNPs with significant bias within a 150 bp rolling window 
# of each other and defined the lead SNP of ASBs by the lowest p value, integrating both read depth and allele frequency bias. 
if(b.disequilibriumLinkageHandling){ # TODO: select to do linkage selection by distance to peak (instead of p-value)
  message("Processing linkage disequilibrium of significant bQTLs...")
  source("utils/linkage_pvalue.R")
}
n_snps_linkage_filtered = nrow(l.selection[[1]])
message(n_snps_linkage_filtered, " SNPs located within high-confidence BZR1 binding peaks within linkage groups")


# Figure: Allelic and spatial distribution of allele-specific BZR1 binding sites with significant bias towards B73 (green)
# or Mo17 (blue) along the maize B73 reference genome. Allelic bias is expressed as percentage
# of B73 read counts. Chromosome borders and length are depicted by dashed lines and arrows at
# the bottom, respectively. Centromere locations are indicated by orange rectangles.
bQTL_scatterplot(postTotal=l.selection[[1]]) # fig 2 c
df.ASBs <- l.selection[[1]]


message("perform gene partitioning")
df.ASB_gene_partitioning = add_genomic_location(df.ASBs,  
                                                pos_peak = "position" ,
                                                chr_Nr = "contig",
                                                df.gene_annotation,
                                                v.genePartitions = c("gene", "five_prime_UTR",  "CDS", "three_prime_UTR", "exon"),
                                                n.cpus = 1)

message(nrow(df.ASB_gene_partitioning), " high confidence allele-specific BZR1 binding sites located near ", length(unique(df.ASB_gene_partitioning$gene.ID)), " flanking genes")

df.distribution <- data.frame(ASBs = apply(df.ASB_gene_partitioning[,v.partitions], 2, table)["yes",])
df.distribution["%ASBs"] <-  round(df.distribution$ASBs / sum(df.distribution$ASBs) * 100, 1)

message("Distribution of ASBs positions around / withing genes:")
print(df.distribution)



df.peaks_gene_partitioning = add_genomic_location(df.peaks, 
                                                  pos_peak = "posPeak" ,
                                                  chr_Nr = "seqnames",
                                                  df.gene_annotation,
                                                  v.genePartitions = c("gene", "five_prime_UTR",  "CDS", "three_prime_UTR", "exon"),
                                                  n.cpus = 1)

message(nrow(df.peaks_gene_partitioning), " high confidence allele-specific BZR1 binding peaks located near ", length(unique(df.peaks_gene_partitioning$gene.ID)), " flanking genes")


if(b.write_paper){
  write.table(df.peaks_gene_partitioning, paste(folder_output, "/S4.txt", sep = ""),  row.names = FALSE, quote = FALSE, sep ="\t")
}



# TODO: identify background gene partitioning - after sampling  message("Figure 2 - Relative comparative partition analysis ")

# TODO: create function from gene partition - 
write.csv(df.partitions, paste("output/df.partition_plus_", timeStamp, ".csv", sep = ""))

# TODO: background sampling -> gene partitioning alignment 




message("Control background SNP (bgSNP) sampling")
# Functional GWAS variants have been shown to be significantly enriched in gene proximal
# regions. Therefore, control background SNPs (bgSNPs) were proportionally sampled (excluding ASBs) 
# per chromosome and genomic location (i.e. 5 - 1kb upstream, 1kb upstream - TSS, 5’UTR, exon, intron, 3’UTR, TTS - 1 kb downstream, intergenic) to match
# the genomic distribution of the ASBs dataset. Additionally, we checked that ASBs and bgSNPs showed a similar minor allele-frequency . In total,
# 168950 bgSNPs were sampled, yielding approximately 50 times as many background SNPs per genome location, compared to the number of ASBs within each location

if(!b.load_snp_to_gene_partitioning_and_background_sampling){
  
  source("utils/HashSeq_populationSampling.R")
  l.postTotal = sample_nonsignificant_background_snps_in_peaks(l.postTotal)
  
  
  # run gene partitioning 
  # run the background and gene partitioning 
  source("utils/HachSeq_partitioning.R")
  
  
  # TODO: input... here would have been ... 
  # l.bQTL_gene_partitioning <- readRDS(paste("tmp/l.bQTL_gene_partitioning_", timeStamp, ".rds", sep = ""))
  l.bQTL_gene_partitioning = perform_SNP_to_gene_partitioning(df.gene_annotation=df.gene_annotation,
                                                              v.genePartitions=v.genePartitions)
  
  l.res <- create_background_distribution_with_similar_gene_partitioning(l.bQTL_gene_partitioning = l.bQTL_gene_partitioning, multiplyer = s.multiplyer, 
                                                                         v.partitions = v.partitions, b.duplicateRemoval = b.duplicateRemoval, seed.randomGenerator = 1234)
  
  l.bQTL_gene_partitioning.sampling <- l.res$l.bQTL_gene_partitioning
  df.distribution <- l.res$df.distribution
  
  # saveRDS(l.res, paste(folder_tmp,"l.bQTL_gene_partitioning_withGeneDistances_backgroundSampled.rds", sep="/")) 
  
  if(b.save_intermediate_results){
    
    #saveRDS(l.bQTL_gene_partitioning.sampling, paste("tmp/l.bQTL_gene_partitioning_no_duplicates_bg_sampled.", timeStamp, ".rds", sep = ""))  

    #df.bQTL_gene_partitioning_bg_equivalent <- read.csv("output/df.bQTL_gene_partitioning_bg_equivalent.csv")
  }
  
}else{
  
  l.bQTL_gene_partitioning <- readRDS(paste(folder_tmp, "l.bQTL_gene_partitioning_withGeneDistances_backgroundSampled.rds", sep = "/"))$l.bQTL_gene_partitioning
  l.bQTL_gene_partitioning[[1]] <- df.ASB_gene_partitioning
  df.bQTL_gene_partitioning_bg_equivalent <- l.bQTL_gene_partitioning[[2]]
  
  #df.distribution <- readRDS(paste(folder_tmp,"l.bQTL_gene_partitioning_withGeneDistances_backgroundSampled.rds", sep="/"))$df.distribution

}

df.distribution["bgSNPs"] <- apply(df.bQTL_gene_partitioning_bg_equivalent[,v.partitions], 2, table)["yes",]
df.distribution["%bgSNPs"] <-  round(df.distribution$bgSNPs / sum(df.distribution$bgSNPs) * 100, 1)
df.distribution["multiplyer"] <- df.distribution$bgSNPs / df.distribution$ASBs

message("Distribution of ASBs and background Snps positions around / withing genes:")
print(df.distribution)


if(b.write_paper){ # TODO: important - exchange with df.ASBs
  
  write.csv(df.distribution, paste("../output/df.ASB_and_bgSnp_gene_partitions.csv", sep = ""))  
  write.csv(l.bQTL_gene_partitioning.sampling[[1]], paste("output/df.ASB_gene_partitioning.csv", sep = ""))  
  write.csv(l.bQTL_gene_partitioning.sampling[[2]], paste("output/df.bgSNP_gene_partitioning.csv", sep = ""))  
  
  write.table(l.bQTL_gene_partitioning[[1]], paste(folder_output, "manuscript/supplement/S8.txt", sep = "/"),  row.names = FALSE, quote = FALSE, sep ="\t")
}
message(nrow(l.bQTL_gene_partitioning[[2]]), " bgSNPs have been sampled")





message("Genomic feature profiling of ASBs (Methylation, Motifs and DNAseI, Enhancers) ")



message("------------ Methylation ---------")
# Methylation levels for CG, CHG, and CHH for B73 and Mo17 were extracted from Regulski et al. 2013
# Methylation frequency versus distance (up to +/- 2 kbp) around
# each ASB were averaged over 20bp bins, and visualized by regions bound by BZR1 with
# either high or low affinity levels depending on the inbred line. For B73, high and low
# affinity bound regions were defined by a post frequency of >= 0.85 or <= 0.15, respectively
# and oppositely for Mo17 by a post frequency <= 0.15 and >= 0.85), respectively.

source("utils/methylation.R")
# asb_distance_vs_methylation(df.ASBs)


# df.ASBs # readRDS(paste("tmp/l.bQTL_gene_partitioning_withGeneDistances_backgroundSampled_", timeStamp, ".rds", sep = ""))[[1]]


# To identify ASBs which overlapped with significant variation in either CG, CHG, or CHH
# methylation between B73 and Mo17, we first, per ASB, assigned averaged methylation
# levels of Mo17 and B73 methylation levels (separately for the CG, CHG, or CHH
#  methylation datasets) within a given window of +/- 40 bp around the ASB position.
# Differentially methylated alleles were defined as described previously 13 . Accordingly, we
# identified ASBs as overlapping with differentially methylated regions if in the B73 or Mo17
# methylation datasets, the methylation level of one allele would be >= 70% while the level
# of the corresponding allele was <= 10%.





# TODO: create subdirectories for plots !!!
methylation_occupancy(l.bQTL_gene_partitioning, 
                      th.distance_to_ASB = 2000,
                      n.chromosomes = 10,
                      n.cpus = 4,
                      n.binWidth = 20,
                      b.val = TRUE,
                      v.species = c("Mo17", "B73"),
                      v.groups = c("all", "genic", "non_genic"),
                      v.species = c("Mo17", "Mo17" , "B73", "B73"),
                      v.stringency <- c("< 0.5 | > 0.5", "< 0.15 | > 0.85", "< 0.5 | > 0.5", "< 0.15 | > 0.85"),
                      v.filenames = c(paste(folder_data,"methylation/B73_CHG.bw", sep = "/"),
                                      paste(folder_data,"methylation/B73_CHH.bw", sep = "/"),
                                      paste(folder_data,"methylation/B73_CpG.bw", sep = "/"),
                                      paste(folder_data,"methylation/MO17_CHG.bw", sep = "/"),
                                      paste(folder_data,"methylation/MO17_CHH.bw", sep = "/"),
                                      paste(folder_data,"methylation/MO17_CpG.bw", sep = "/")),
                      v.datasets = c("B73_CHG", "B73_CHH", "B73_CpG", "MO17_CHG", "MO17_CHH", "MO17_CpG"),
                      v.formats = c("bigWig", "bigWig", "bigWig", "bigWig", "bigWig", "bigWig"),
                      folder_tmp = folder_tmp,
                      folder_output = folder_output,
                      do.plot = F)


# rename methylation, open chromatin?
l.SNPs_w_chromatin <- RDS(paste(folder_tmp, "l.SNPs_w_chromatin_ASBs_and_bgSNPs.rds", sep = "/"))

# Fig. 4 l) Correlation of CG methylation with allele-specific BZR1 binding. Average
# CG methylation of the B73 - Mo17 allele of the 40bp surrounding each ASB are plotted against
# the allelic bias (expressed in percentage of B73 read counts). Differential CG methylation is
# indicated by red dots.
differential_methylation_vs_allelic_bias(l.SNPs=l.bQTL_gene_partitioning,
                                        th.bp_offset = 20,
                                        th.distance_to_ASB = 2000,
                                        n.chromosomes = 10,
                                        n.cpus = 1,
                                        b.val = TRUE,
                                        v.species = c("Mo17", "B73"),
                                        degrees = 15,
                                        th.distance_to_ASB = 5000,
                                        th.padding = 50,
                                        width = 6,
                                        height = 4,
                                        group = "genic_and_nongenic",
                                        v.filenames = c("data/dnase/GSE94291_DNase_ist.bedGraph", 
                                                        "data/methylation/B73_CpG.bw", 
                                                        "data/methylation/MO17_CpG.bw",
                                                        "data/methylation/B73_CHG.bw",
                                                        "data/methylation/MO17_CHG.bw",
                                                        "data/methylation/B73_CHH.bw",
                                                        "data/methylation/MO17_CHH.bw"),
                                        v.datasets = c("DNase", "B73_CpG",  "MO17_CpG", "B73_CHG", "MO17_CHG", "B73_CHH", "MO17_CHH"),
                                        v.formats = c("bedGraph", "bigWig", "bigWig", "bigWig", "bigWig", "bigWig", "bigWig"),
                                        v.has_ranges = c("yes",  "no", "no",  "no", "no",  "no", "no")
)


# TODO: get absolute and differential figure ... 


l.SNPs_w_chromatin <- readRDS(paste(folder_tmp, "l.SNPs_w_chromatin_ASBs_and_bgSNPs.rds", sep = "/"))


df.ASBs



l.SNPs=l.bQTL_gene_partitioning
th.bp_offset = 20
th.distance_to_ASB = 2000
n.chromosomes = 10
n.cpus = 1
b.val = TRUE
v.species = c("Mo17", "B73")
degrees = 15
th.distance_to_ASB = 5000
th.padding = 50
width = 6
height = 4
group = "genic_and_nongenic"
v.filenames = c("data/dnase/GSE94291_DNase_ist.bedGraph", 
                "data/methylation/B73_CpG.bw", 
                "data/methylation/MO17_CpG.bw",
                "data/methylation/B73_CHG.bw",
                "data/methylation/MO17_CHG.bw",
                "data/methylation/B73_CHH.bw",
                "data/methylation/MO17_CHH.bw")
v.datasets = c("DNase", "B73_CpG",  "MO17_CpG", "B73_CHG", "MO17_CHG", "B73_CHH", "MO17_CHH")
v.formats = c("bedGraph", "bigWig", "bigWig", "bigWig", "bigWig", "bigWig", "bigWig")
v.has_ranges = c("yes",  "no", "no",  "no", "no",  "no", "no")


# 
# Fig. 2. Variation in DNA sequence and methylation underlie differential BZR1 binding. a)
# Schematics of HASCh-seq approach and possible causes for allele-specific binding events.
# Chromatin-IP is performed in F1 hybrid plants. From top to bottom, three possible scenarios for
# TF binding to the parental genomes (green and blue) are depicted, where binding strength is
# represented by the line width of the black arrows: when no alterations in motif and chromatin
# structure occurs, binding is expected to be equal. Lower binding is expected If the binding motif
# is altered by a SNP or epigenetic features like DNA methylation for one of the alleles. b) An
# example of allele-specific binding of ZmBZR1 to target gene Zm00001d031434. BZR1 bound
# target reads that uniquely map to either B73 (green) or Mo17 (blue) are shown. c) Allelic and
# spatial distribution of allele-specific BZR1 binding sites with significant bias towards B73 (green)
# or Mo17 (blue) along the maize B73 reference genome. Allelic bias is expressed as percentage
# of B73 read counts. Chromosome borders and length are depicted by dashed lines and arrows at
# the bottom, respectively. Centromere locations are indicated by orange rectangles. A red box on
# 26chromosome 1 marks the ASB upstream of Zm00001d031434 displayed in Fig. 2b. d) Genomic
# distribution of ASBs. 3297 ASBs were classified according to their location relative to genes. In
# case of two genes in the proximity of an ASB, the priority given was exon>intron>UTR>1 kb
# upstream>1 kb downstream>1-5 kb upstream. e) Frequency of BRRE (CGTG[C/T]G), G-Box
# (CACGTG), and the control motif GTACGG (SBP-box 12 ) around ASBs of the alleles with higher
# BZR1 binding. f) Fraction of ASBs overlapping with motifs, for which the allele with canonical
# BRRE, G-Box or a control motif GCCGCC (GCC-box 12 ), showed higher ZmBZR1 affinity. Both
# BR-related BRRE and G-box motifs, but not the control motif, diverge significantly (p<0.001,
# Fisher’s exact test) from the expected 50% random distribution. g) Distribution of lead SNPs of
# ASBs within motifs. Among ASBs which overlapped with BRRE motif, SNPs were enriched in the
# core CG bases. h) ASBs affecting BZR1 motifs and/or overlapping with differentially methylated
# (CpG, CHG or CHH) sites between B73 and Mo17. i)-k) Average CpG (h), CHG (i) and CHH(j)
# methylation frequency in B73 (green) and Mo17 (blue) over ASB loci with a least 85% binding
# bias towards B73 or Mo17. High affinity ( ___ ) and low affinity ( …. ) alleles are considered separately
# for each genotype. l) Correlation of CG methylation with allele-specific BZR1 binding. Average
# CG methylation of the B73 - Mo17 allele of the 40bp surrounding each ASB are plotted against
# the allelic bias (expressed in percentage of B73 read counts). Differential CG methylation is
# indicated by red dots.




message("------------ Motifs ---------")
# To identify ASBs which may be explained by motif variation (Fig. 2h), we extracted the
# +/- 5 bp of the high affinity BZR1 bound allele surrounding ASBs. Using R we scanned
# those 11 bp fragments for canonical BRRE (CGTG[T/C]G, C[G/A]CACG), allowing a
# single base pair mismatch outside the core motif (CGTG), or G-box (CACGTG) motifs
# and determined ASBs where the SNP changed a BRRE or G-box motif into an altered
# (non BRRE or G-box) motif. 

if(!b.load_motif_analysis){
  source("motifs/predefined_motif_analysis.R")
  predefined_motif_analysis(l.bQTL_gene_partitioning, df.peaks, motifs, v.motif_offset)
  
  l.motif_analysis <- readRDS(paste(folder_tmp, "l.motif_analysis.rds", sep = "/"))
  l.nucleotideInPeaks <- readRDS(paste(folder_tmp, "l.nucleotideInPeaks.rds", sep = "/"))
  
  
  source("motifs/predefined_motifs_in_peaks.R")
  predefined_motifs_in_peaks(l.motif_analysis, l.nucleotideInPeaks)
  
  l.motif_analysis.postprocessed = readRDS(paste(folder_tmp, "l.motif_analysis.postprocessed.rds", sep = "/"))
  
  
  source("motifs/directionality_and_snp_distribution.R")
  directionality_and_snp_distribution(l.motif_analysis.postprocessed)
  
  df.motif_nucleotidePositionPartitions <- readRDS(paste(folder_tmp, "df.motif_nucleotidePositionPartitions.rds", sep = "/"))
  
  
  source("motifs/motif_snp_position_significance.R")
  motif_snp_position_significance(l.bQTL_gene_partitioning, motifs, v.motif_offset, df.peaks, genome, l.genome.mutant, n.chromosomes)
  

  source("motifs/evaluating_BZR1_core_motif.R")
  evaluating_BZR1_core_motif(df.ASBs, core_motif = "CGTG", s.motif_offset = 5, n.chromosomes = 10)
  
}else{
  l.motif_analysis <- readRDS(paste(folder_tmp, "l.motif_analysis.rds", sep = "/"))
  l.nucleotideInPeaks <- readRDS(paste(folder_tmp, "l.nucleotideInPeaks.rds", sep = "/"))
  l.motif_analysis.postprocessed = readRDS(paste(folder_tmp, "l.motif_analysis.postprocessed.rds", sep = "/"))
  df.motif_nucleotidePositionPartitions <- readRDS(paste(folder_tmp, "df.motif_nucleotidePositionPartitions.rds", sep = "/"))
  df.motifPositionAnalysis <- readRDS(paste(folder_tmp, "df.motifPositionAnalysis.rds", sep = "/"))
  df.ASBs.core_motifs <- readRDS(paste(folder_tmp, "df.ASBs.core_motifs.rds", sep = "/"))
}

if(b.write_paper){
  write.table(l.motif_analysis[[1]], paste(folder_output, "/motifs/S0_1.txt", sep = ""),  row.names = FALSE, quote = FALSE, sep ="\t")
  
  print(table(l.motif_analysis.postprocessed[[1]]$direction))
  write.table(l.motif_analysis.postprocessed[[1]], paste(folder_output, "/motifs/S0_2.txt", sep = ""),  row.names = FALSE, quote = FALSE, sep ="\t")
  write.table(df.motif_directionality, paste(folder_output, "/motifs/S0_3.txt", sep = ""),  row.names = FALSE, quote = FALSE, sep ="\t")
  write.table(df.motif_nucleotidePositionPartitions, paste(folder_output, "/motifs/S0_4.txt", sep = ""),  row.names = FALSE, quote = FALSE, sep ="\t")
  write.table(df.motifPositionAnalysis, paste(folder_output, "/motifs/S0_5.txt", sep = ""),  row.names = FALSE, quote = FALSE, sep ="\t")
  write.table(df.ASBs.core_motifs, paste(folder_output, "/SX3.txt", sep = ""), quote = FALSE, row.names = FALSE, sep ="\t")
  
}



message("------------ Enhancers ---------")
# In addition to cis-regulatory elements located in the close vicinity to the transcriptin
# start site (TSS) or transcription termination site (TTS), transcriptional enhancers can be
# located tens of kilobases from their target genes 28 . We thus hypothesized that the ASBs
# located in intergenic regions could be located within distant enhancers and analyzed their
# abundance around putative intergenic enhancer regions 29 . Of the ASBs distant from the
# genes, 11% coincided with the 1495 intergenic enhancer regions identified in B73, which
# is approximately 37-fold higher than random probability (Fig. 4a).

# Putative B73 enhancer (Fig. 4a) regions were extracted from Oka et al. 2017 15 . To
# determine potential enrichments, +/- 10 kbps surrounding enhancer regions were
# intersected with ASBs and bgSNPs.

source("utils/enhancer.R")

n.maxDist = 100000 # TODO: discuss - currently 100000 not 10000 chosen  

res<- snp_to_enhancer_distance(l.bQTL_gene_partitioning,
                              df.enhancer_genes,
                              v.sets = c("ASB", "bgSNP"),
                              n.binWidth = 10,
                              n.maxDist = n.maxDist)

# Fig. 4. a) ASBs of ZmBZR1 are overrepresented at enhancer sites as compared to randomly selected background (bg) SNPs
if(b.write_paper){
  pdf(paste(folder_output, paste("enhancer/ASBs_of_ZmBZR1_overrepresented_at_enhancer_sites.pdf", sep = ""), sep ="/"), width = 6, height = 4)
  print(res$p.dist_density)
  dev.off()
}

message("RNA extraction, RNA sequencing and differential expression analysis")
# BR deficient brd1 mutant siblings were grown in soil under greenhouse conditions for 26
# days as described above. The oldest 2 leaves were removed and 2 cm of meristem
# enriched tissue (Supplementary Fig. 1) were placed in 1 µM BL for 4 h at room
# temperature (RT) in water. Total RNA was isolated using acidic phenol extraction as
# described previously10. Purification of poly-adenylated mRNA using oligo(dT) beads,
# construction of barcoded libraries, and sequencing using Illumina HiSeq 4500 technology
# (75 bp paired-end reads) were performed by Novogene Co. using manufacturer's Illumina
# recommendations. Trimmed and QC (Seqpurge v. 2019-02-11) filtered sequence reads
# were mapped to the B73 AGPv4 genome using STAR (v. 2.54)11 in twopass mode (with
# parameters: --outFilterScoreMinOverLread 0.3, --outFilterMatchNminOverLread 0.3, --
# outSAMstrandField intronMotif, --outFilterType BySJout, --outFilterIntronMotifs
# RemoveNoncanonical, --quantMode TranscriptomeSAM GeneCounts). Unique reads
# were filtered by mapping quality (q20) and PCR duplicates removed using Samtools (v.
# 1.3.1). Gene expression was analyzed in R (v. 3.4.1) using the DEseq2 software (v.
# 1.16.1)12. Genes were defined as differentially expressed by a 1.5-fold expression
# difference with a p value, adjusted for multiple testing, of < 0.05. Accession number for
# the RNA-seq data in the Gene Expression Omnibus database will be available upon
# paper acceptance. For the analysis of allele-specific gene regulation a previously
# published RNA-seq data set from B73xMo17 F1 hybrids was used (Baldauf et al., 2016).
source("expression/get_brassinolide_treatment_rnaseq_data.R")
df.bQTL_RNAseq <- get_brassinolide_treatment_rnaseq_data(path.rnaseq.down_regulated, path.rnaseq.up_regulated)
message(nrow(df.bQTL_RNAseq), " BR-responsive genes ( repressed ", table(df.bQTL_RNAseq$mode)[1]," , activated ", table(df.bQTL_RNAseq$mode)[2], " )")
if(b.write_paper){
  write.table(df.bQTL_RNAseq, paste(folder_output, "/S1.txt", sep = ""), quote = FALSE, row.names = FALSE, sep = "\t")
}





message("-------------- ChIP-seq ------------- ")
# Quality filtered ChIP-seq reads were aligned to the B73 AGPv4 genome using bwa-mem (v. 0.7.16a)
# with default parameters, followed by removal of PCR duplicates using samtools (v. 1.3.1.). 
# To determine BZR1 binding peaks, IP and negative control samples,
# after normalization for read depth, were analyzed using the GEM package (v. 3.0) (using
#   parameters: --fold 5, --k_min 5, --k_max 8). After samples were analyzed individually,
# peaks reproducible in all 3 replicas, using a 400 bp window (+/- 200 bp around each peak)
# were determined using R (v. 3.3.2) and considered high confidence peaks. 
# Our Chip-seq experiment identified 17463 high confidence ZmBZR1 binding peaks

if(!b.load_chip_analysis){
  source("utils/chipseq_binding_peaks.R")
  chipseq_binding_peaks(l.path.ChipSeqB73, s.half_window_size.input, n.chromosomes)
  
  df.ChipSeq.filtered <- readRDS(paste(folder_tmp,"df.ChipSeq.filtered.rds", sep ="/"))
  
  # The ZmBZR1 ChIP-seq peaks located near 6371 putative target genes (Supplementary Table 3)
  source("utils/genomic_location.R")
  df.ChipSeq_gene_partitioning = add_genomic_location(df.ChipSeq.filtered, 
                                                      pos_peak = "pos" ,
                                                      chr_Nr = "chr",
                                                      df.gene_annotation,
                                                      v.genePartitions = c("gene", "five_prime_UTR",  "CDS", "three_prime_UTR", "exon"),
                                                      n.cpus = 1)
  
  saveRDS(df.ChipSeq_gene_partitioning, paste(folder_tmp,"df.ChipSeq_gene_partitioning.rds", sep ="/"))
  
}else{
  df.ChipSeq.filtered <- readRDS(paste(folder_tmp,"df.ChipSeq.filtered.rds", sep ="/"))
  df.ChipSeq_gene_partitioning <- readRDS(paste(folder_tmp,"df.ChipSeq_gene_partitioning.rds", sep ="/"))
}

if(b.write_paper){
  write.table(df.ChipSeq.filtered, paste(folder_output, "chipseq/S2.txt", sep = "/"),  row.names = FALSE, quote = FALSE, sep ="\t")
  write.table(df.ChipSeq_gene_partitioning, paste(folder_output, "/S3.txt", sep = ""),  row.names = FALSE, quote = FALSE, sep ="\t")
}

v.chipseq_target_genes <- unique(df.ChipSeq_gene_partitioning$gene.ID)
v.chipseq_target_genes <- v.chipseq_target_genes[!is.na(v.chipseq_target_genes)]
message(nrow(df.ChipSeq.filtered), " high confidence ZmBZR1 ChiP-seq binding peaks located near ", length(v.chipseq_target_genes), " putative target genes of ZmBZR1")


if(!b.load_chip_analysis){

  # The ZmBZR1 ChIP-seq peaks located near 6371 putative target genes, which included (according to the RNA-seq analysis) 
  # 580 and 469 BR activated and repressed genes, respectively
  df.bQTL_gene_partitioning <- merge(df.ChipSeq_gene_partitioning, df.bQTL_RNAseq, by = "gene.ID")
  df.bQTL_gene_partitioning <- subset(df.bQTL_gene_partitioning, df.bQTL_gene_partitioning$mode %in% c("up", "down"))
  df.gene_set <- unique(df.bQTL_gene_partitioning[, c("gene.ID", "mode")])
  message(nrow(df.gene_set), " BR-responsive putative target genes of ZmBZR1 ( repressed ", table(df.gene_set$mode)[1]," , activated ", table(df.gene_set$mode)[2], " )")

  source("utils/gene_features.R")
  df.ChipSeq_gene_partitioning <- add_gene_features(df.ChipSeq_gene_partitioning,
                                                    df.gene_function,
                                                    df.gene_orthologs,
                                                    df.gene_conversion.AGPv3_to_AGPv4=df.geneID_conversion)
  
  df.ChipSeq_gene_partitioning <- subset(df.ChipSeq_gene_partitioning, !is.na(df.ChipSeq_gene_partitioning$Arabidopsis_ortholog))
  df.ChipSeq_gene_partitioning <- subset(df.ChipSeq_gene_partitioning, df.ChipSeq_gene_partitioning$Arabidopsis_ortholog != "")
  df.ChipSeq_gene_partitioning$Arabidopsis_ortholog <- gsub("\\..*","", df.ChipSeq_gene_partitioning$Arabidopsis_ortholog)
  df.ChipSeq_gene_partitioning = subset(df.ChipSeq_gene_partitioning, df.ChipSeq_gene_partitioning$non_genic == "no")
  message(length(unique(df.ChipSeq_gene_partitioning$gene.ID)), " BR-responsive putative target genes of ZmBZR1 with ", length(unique(df.ChipSeq_gene_partitioning$Arabidopsis_ortholog)), " Arabidopsis orthologs" )
  
}else{
  
  
  
}

# message(nrow(df.bQTL_RNAseq), " BR-responsive genes ( repressed ", table(df.bQTL_RNAseq$mode)[1]," , activated ", table(df.bQTL_RNAseq$mode)[2], " )")
message(length(unique(df.ChipSeq_gene_partitioning$gene.ID)), " BR-responsive putative target genes of ZmBZR1 with ", length(unique(df.ChipSeq_gene_partitioning$Arabidopsis_ortholog)), " Arabidopsis orthologs" )


if(b.write_paper){
  write.table(df.ChipSeq_gene_partitioning, paste(folder_output, "/S4.txt", sep = ""), quote = FALSE, row.names = FALSE, sep = "\t")
}


source("utils/br_vs_non_br_regulated.R")
arabidopsis_chipseq_venn_overlap(df.ChipSeq_gene_partitioning)



# TODO: check data online 


message("-------------- Arabidopsis ortholog comparative analysis ------------- ")

source("utils/gene_features.R")
df.ASB_gene_partitioning <- add_gene_features(df.ASB_gene_partitioning,
                                              df.gene_function,
                                              df.gene_orthologs,
                                              df.gene_conversion.AGPv3_to_AGPv4=df.geneID_conversion)

df.ASB_gene_partitioning <- subset(df.ASB_gene_partitioning, !is.na(df.ASB_gene_partitioning$Arabidopsis_ortholog))
df.ASB_gene_partitioning <- subset(df.ASB_gene_partitioning, df.ASB_gene_partitioning$Arabidopsis_ortholog != "")
df.ASB_gene_partitioning$Arabidopsis_ortholog <- gsub("\\..*","", df.ASB_gene_partitioning$Arabidopsis_ortholog)
df.ASB_gene_partitioning = subset(df.ASB_gene_partitioning, df.ASB_gene_partitioning$non_genic == "no")


# BR vs non BR regulation s
source("utils/br_vs_non_br_regulated.R")
br_vs_non_br_regulated(filename = "data/arabidopsis_overlap_genelists/ZmvsAth_1_to_1.txt")




# not identical to figure 1 h (check different variations)
bQTL_on_ArabidopsisHomolog_geneexpression_evaluation <- function(df.bQTL_RNAseq, 
                                                                 path.gene_orthologs = "data/GeneAnnotation/Phytozome13_anno_1_1.txt")



# TODO: where are these datasets
# ath chip seq - dark
# df.At_BZR1_ChipSeqTargets <- read.table("data/ArabidopsisScriptsAndDatasets/At_BZR1_ChipSeqTargets.txt", header = TRUE, sep ="\t", quote = "", stringsAsFactors = FALSE)
# names(df.At_BZR1_ChipSeqTargets) <- "locus"
# 
# # ath chIP chip - light
# df.At_BZR1_ChIPchipTargets <- read.table("data/ArabidopsisScriptsAndDatasets/At_BZR1_ChIPchip.txt", header = TRUE, sep ="\t", quote = "", stringsAsFactors = FALSE)
# 
# length(intersect(df.ASB_gene_partitioning$Arabidopsis_ortholog, df.At_BZR1_ChipSeqTargets$locus))
# length(intersect(df.ASB_gene_partitioning$Arabidopsis_ortholog, df.At_BZR1_ChIPchipTargets$locus))

df.bQTL_RNAseq



# else{
#   message("no denovo motif and in peak analysis")
#   if(b.BRRE_GBOX.only){
#     l.motif_analysis <- readRDS(paste("tmp/l.motif_analysis.BRRE_GBOX_only_",timeStamp, ".rds", sep = ""))
#     l.nucleotideInPeaks <- readRDS(paste("tmp/l.nucleotideInPeaks.BRRE_GBOX_only_",timeStamp, ".rds", sep = ""))
#   }else{
#     l.motif_analysis <- readRDS(paste("tmp/l.motif_analysis_",timeStamp, ".rds", sep = ""))
#     l.nucleotideInPeaks <- readRDS(paste("tmp/l.nucleotideInPeaks_",timeStamp, ".rds", sep = ""))
#   }
# }

write.csv(df.bQTL_gene_partitioning, "output/df.bQTL_gene_partitioning_peaks.csv", row.names = FALSE)


### orthologs ... 

df.bQTL_gene_partitioning.subset <- subset(df.bQTL_gene_partitioning, df.bQTL_gene_partitioning$nonGenic == "no")
length(unique(df.bQTL_gene_partitioning.subset$gene.ID))

length(unique(df.bQTL_gene_partitioning.subset$Arabidopsis_ortholog))





message("-------------- GWAS ------------- ")





# Fig. 4. ASBs of ZmBZR1 are linked to growth and disease related traits. 
# b) Association of ASBs with the curated 4041 significant GWAS hits for selected phenotypes of
# the NAM population. Abbreviations: Av: Average; intern: internode; len: length; w. pl.: whole plant;
# b.: below c) VCAP Variance component analysis results. Variance explained (h2) by the ASB-
#   SNP set (bars) and background SNP set (violin plots, derived from the permutation results).
# 29Orange color in the bars denotes a significantly higher variance explained (h2) by ASBs than
# expected by chance (one-sided permutation test < 0.1).


source("gwas/gwas_enrichment.R")
gwas_enrichment(l.bQTL_gene_partitioning, df.phenotype_gwas, output_file, s.dist_ASB_to_GWAS=2000)



